### PR TITLE
store/tikv: Randomize selection of proxy store on network partition

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Problem Summary: Previously, when TiDB finds a request need to be forwarded, it simply iterates over all the peers sequentially until finding one. However, if there's only 3 TiKV nodes in the cluster and the regions are rarely scheduled, most of their peer lists may have the same order, causing the selected proxy peer very likely to be imbalanced.  This PR tries to randomize the first choice of the proxy peer, and iterate over the list for further retires.

### What is changed and how it works?

What's Changed: Randomize selection of proxy store in GetTiKVRPCContext and OnSendFail.

How it Works:

### Related changes

- Need to cherry-pick to the release branch
  - 5.0

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
  - Before: 
     image
  - After: 
     image

### Release note 

- No release note

### CODE REVIEW

File: store/tikv/region_cache.go [hunk 1]
```diff
@@ -1242,12 +1243,30 @@ func (c *RegionCache) getProxyStore(region *Region, store *Store, rs *RegionStor
 	}
 
 	tikvNum := rs.accessStoreNum(TiKVOnly)
-	for index := 0; index < tikvNum; index++ {
+	if tikvNum <= 1 {
+		return
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1247]: Does it mean to query the leader?

File: store/tikv/region_cache.go [hunk 2]
```diff
@@ -1242,12 +1243,30 @@ func (c *RegionCache) getProxyStore(region *Region, store *Store, rs *RegionStor
...
 		// Skip work store which is the actual store to be accessed
 		if index == int(workStoreIdx) {
 			continue
 		}
 		storeIdx, store := rs.accessStore(TiKVOnly, AccessIndex(index))
+		// Skip unreachable stores.
+		if atomic.LoadInt32(&store.needForwarding) != 0 {
+			continue
+		}
+
```
Comment (youjiali1995) [inline store/tikv/region_cache.go:1269]: I think this function shouldn't report errors. If it reports, the error is returned to the user directly.

File: store/tikv/region_cache.go (hunk 1 above)
Comment (MyonKeminta) [reply store/tikv/region_cache.go:1244]: `GetTiKVRPCContext` will get nil as the proxyStore, so no proxy will be used.

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**